src/caffe/parallel.cpp

```diff
@@ -82,14 +83,17 @@ GPUParams<Dtype>::GPUParams(shared_ptr<Solver<Dtype> > root_solver, int device)
 
   // Allocate device buffers
   CUDA_CHECK(cudaSetDevice(device));
-  CUDA_CHECK(cudaMalloc(&data_, size_ * sizeof(Dtype)));
+  buffer_device_ = device;
+  GPUMemoryManager::allocate(reinterpret_cast<void **>(&data_),
```

Comment:
@longjon is confused because the `cudaMalloc` call that is being replaced doesn't have the ugly `reinterpret_cast`, this is because in `cuda_runtime.h`, `cudaMalloc` has a template wrapper:
```
template<class T>
static __inline__ __host__ cudaError_t cudaMalloc(
  T      **devPtr,
  size_t   size
)
{ 
  return ::cudaMalloc((void**)(void*)devPtr, size);
}
```
If @longjon wants to have an API closer to `cudaMalloc`, we might need to the same.